### PR TITLE
Format message statuses with date if possible.

### PR DESCRIPTION
**Pull request by charlesmchen**
PTAL @michaelkirk 

screen shot 2017-09-27 at 3 20 48 pm

screen shot 2017-09-27 at 3 20 17 pm

### CODE REVIEW

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m [hunk 1]
```diff
@@ -2151,22 +2152,22 @@ - (NSAttributedString *)collectionView:(JSQMessagesCollectionView *)collectionVi
...
-                        }];
-                NSAttributedString *spacing = [[NSAttributedString alloc] initWithString:@" "];
-                result = [[checkmark rtlSafeAppend:spacing referenceView:self.view] rtlSafeAppend:result
-                                                                                    referenceView:self.view];
+            NSString *text;
+            if (outgoingMessage.wasDelivered) {
+                NSNumber *_Nullable firstRecipientReadTimestamp = [outgoingMessage firstRecipientReadTimestamp];
+                if ([OWSReadReceiptManager.sharedManager areReadReceiptsEnabled] && firstRecipientReadTimestamp) {
+                    text = NSLocalizedString(@"MESSAGE_STATUS_READ", @"message footer for read messages");
+                    text = [text rtlSafeAppend:@" " referenceView:self.view];
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: I'm leaning toward using `rtlSafeAppend()` everywhere we can, e.g. situations where we don't anticipate the localization to affect ordering beyond LTR-v.-RTL.  This should significantly simplify the job for our translators.

File: Signal/src/util/NSAttributedString+OWS.h [hunk 2]
```diff
@@ -4,6 +4,12 @@
 
 NS_ASSUME_NONNULL_BEGIN
 
+@interface NSString (OWS)
+
+- (NSString *)rtlSafeAppend:(NSString *)string referenceView:(UIView *)referenceView;
+
+@end
```
Comment (charlesmchen) [inline Signal/src/util/NSAttributedString+OWS.h]: I skipped creating a separate source file for this category.

File: Signal/src/util/DateUtil.m [hunk 3]
```diff
@@ -53,4 +53,36 @@ + (BOOL)dateIsToday:(NSDate *)date {
...
+
+    uint64_t nowTimestamp = [NSDate ows_millisecondTimeStamp];
+    if (pastTimestamp >= nowTimestamp) {
+        OWSCFail(@"%@ Unexpected timestamp", self.tag);
+        return NSLocalizedString(@"TIME_NOW", @"Indicates that the event happened now.");
+    }
+
+    NSDate *pastDate = [NSDate ows_dateWithMillisecondsSince1970:pastTimestamp];
+    if ([self dateIsToday:pastDate]) {
+        return [[self timeFormatter] stringFromDate:pastDate];
```
Comment (michaelkirk) [inline Signal/src/util/DateUtil.m]: nit: should we dispatchOnce all the formatters for perf?

File: Signal/src/util/NSAttributedString+OWS.h (hunk 2 above)
Comment (michaelkirk) [reply Signal/src/util/NSAttributedString+OWS.h]: nit: can you move this to the existing `NSString+OWS.h`?

Comment (charlesmchen) [reply Signal/src/util/NSAttributedString+OWS.h]: Oh good point!

File: Signal/src/util/DateUtil.h [hunk 4]
```diff
@@ -5,10 +5,11 @@
 @interface DateUtil : NSObject
 
 + (NSDateFormatter *)dateFormatter;
-+ (NSDateFormatter *)weekdayFormatter;
 + (NSDateFormatter *)timeFormatter;
 + (BOOL)dateIsOlderThanOneDay:(NSDate *)date;
 + (BOOL)dateIsOlderThanOneWeek:(NSDate *)date;
 + (BOOL)dateIsToday:(NSDate *)date;
 
++ (NSString *)formatPastTimestampRelativeToNow:(uint64_t)pastTimestamp;
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply Signal/src/util/DateUtil.h]: Done.

File: Signal/src/util/DateUtil.m [hunk 5]
```diff
@@ -53,4 +53,36 @@ + (BOOL)dateIsToday:(NSDate *)date {
     return [self date:[NSDate date] isEqualToDateIgnoringTime:date];
 }
 
++ (NSString *)formatPastTimestampRelativeToNow:(uint64_t)pastTimestamp
+{
+    OWSCAssert(pastTimestamp > 0);
+
+    uint64_t nowTimestamp = [NSDate ows_millisecondTimeStamp];
+    if (pastTimestamp >= nowTimestamp) {
+        OWSCFail(@"%@ Unexpected timestamp", self.tag);
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply Signal/src/util/DateUtil.m]: Done.

File: Signal/src/util/DateUtil.m (hunk 3 above)
Comment (charlesmchen) [reply Signal/src/util/DateUtil.m]: Huh, `NSDateFormatter` has been thread-safe since iOS 7.  TIL

**Pull request merged**